packages/grafana-ui/src/types/datasource.ts

```diff
@@ -268,6 +269,9 @@ export abstract class DataSourceApi<
    * in the annotation editor `annotations` capability also needs to be enabled in plugin.json.
    */
   annotationQuery?(options: AnnotationQueryRequest<TQuery>): Promise<AnnotationEvent[]>;
+
+  async getDerivedFields?(data: { [key: string]: string | number | null }): Promise<ScopedVars>;
+  getDataLinks?(context: 'logs' | 'metrics'): DataLink[];
```

Comment:
New API with which the datasource can define Data Links and derived fields.

Reply:
I think @ryantxu and I are thinking of a bit more generic structure for field config rules that can be applied to anything that is on FieldConfig type.